src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/SharedScope.java

```diff
@@ -134,14 +134,15 @@ public boolean isAlive() {
         void add(ResourceCleanup cleanup) {
             while (true) {
                 ResourceCleanup prev = (ResourceCleanup) FST.getAcquire(this);
-                cleanup.next = prev;
-                ResourceCleanup newSegment = (ResourceCleanup) FST.compareAndExchangeRelease(this, prev, cleanup);
-                if (newSegment == ResourceCleanup.CLOSED_LIST) {
+                if (prev == ResourceCleanup.CLOSED_LIST) {
+                    cleanup.next = null; // Release, as long as two threads can't add same cleanup it's ok
                     // too late
                     throw new IllegalStateException("Already closed");
-                } else if (newSegment == prev) {
-                    return; //victory
                 }
+                cleanup.next = prev;
```

Comment:
Why not moving this assigment inside the `if` block (e.g. when add succeeded) and then remove the `cleanup.next = null` ?

Reply:
Do you mean something like this?
```
if (FST.compareAndSet(this, prev, cleanup)) {
cleanup.next = prev;
}
```
It could create race with clean block, as clean block could read and process new head, before the pointer to next has been assigned.

(I actually bit wonder, if this cleaning is required, I thought about case that this class will be kept by someone, and it can create some memory leak, but I think there's no many places where it can happen)

Reply:
I guess what I was suggesting is this:

```
void add(ResourceCleanup cleanup) {
            while (true) {
                ResourceCleanup prev = (ResourceCleanup) FST.getAcquire(this);
                if (prev == ResourceCleanup.CLOSED_LIST) {
                    // too late
                    throw new IllegalStateException("Already closed");
                }
                if (FST.compareAndSet(this, prev, cleanup)) {
                    cleanup.next = prev;
                    return; //victory
                };
                // keep trying
            }
        }
 ```

E.g. setup `prev` only if we are sure that the element has been added to the list - but that probably won't work - as it leaves the added element in a partially invalid state.